### PR TITLE
PFIN-68 - Enhancements for ISS Interface v1.0.3 - new EMV tags 

### PR DESCRIPTION
**Pull request by fsemerdzhieva**
Initial changes to the ISS Service Provider:
CHANGED:
v1.0.2 --> v1.0.3
CashWithDrawalReq:
mcTraceId - new field
transactionTechnology- new enum value ECOMMERCE
acquirerAuth.acquirerId ->acquirerAuth.processingNetwork
acquirerAuth.acquirerReferenceNumber -> acquirerAuth.retrievalReferenceNumber
merchant.merchantCompanyName -> merchant.merchantName
merchant.merchantId new field
terminalData.terminalTrxStatus new field
terminalData.terminalCATLevel new field
terminalData.terminalCapability new field
terminalData.terminalSupportIndicator new field
terminalData.partialApprovalSupported new field default to fasle
terminalData.terminalLocation  new field
terminalData.stan to terminalData.STAN new field
card.PANSequenceNumber new field default 0
card.encryptedPIN -> base64 encoded new field
track2Data -> masked new field
EMVRequestData.transactionTime new field
change some of the fields to base64 encoding (ref swagger)
TO DO from v1.0.2:
EMVRequestData.applicationExpirationDate
EMVRequestData.applicationIdentifier
EMVRequestData.applicationPANSequenceNumber 
EMVRequestData.transactionSequenceCounter
EMVRequestData.track2EquivalentData
Response:
add respomse body into commonResponse* object
amountBillingPartial.moneyAmount.currency new field
amountBillingPartial.moneyAmount.amount new field
amountBillingPartial.currencyRate new field
TO DO from v1.0.2:
EMVResponse.issuerAuthenticationData mapping?? (91)

Reversal Reguest:
amountAuthPartial.currency
amountAuthPartial.amount
amountBillingPartial.moneyAmount.currency
amountBillingPartial.moneyAmount.amount
amountBillingPartial.currencyRate  -> used this to set the txnauthorised amount->this needs to be checked (QAed)
transactionTechnology

OUTSTANDING:
*to be confirmed merchantId
*issuerScriptResults
*EMV Response Data
* Additional Changes Required from Martin

### CODE REVIEW

File: ISS/develop/scripts/Build Request Message.js [hunk 1]
```diff
@@ -187,8 +214,43 @@ function populateTerminalDataObj(request) {
...
+
+    try {
+        var sf61_10 = txn.getMcdmData().getMcdmposdata().getCardholderactivatedterminallevel();
+        terminalDataObj.terminalCATLevel = sf61_10;
+    } catch (ex) {
+        log.debug('DE61sf10 - terminalCATLevel not present in the MCDM message');
+    }
+
+    try {
+        var f48sf61_2 = txn.getMcdmSchemaData().getMcdmposDataExtendedCondCodes().getPurchAmtOnlyTermSuppInd() === 1 ? true : false;
```
Comment (dianavicol) [inline ISS/develop/scripts/Build Request Message.js:234]: `mcdmSchemaData` field doesn't exist, also, `isPurchAmtOnlyTermSuppInd()` could be used

File: ISS/develop/scripts/Build Request Message.js [hunk 2]
```diff
@@ -187,8 +214,43 @@ function populateTerminalDataObj(request) {
...
+
+    try {
+        var sf61_3 = txn.getMcdmData().getMcdmposdata().getPosterminallocation();
+        terminalDataObj.terminalLocation = sf61_3;
+    } catch (ex) {
+        log.debug('DEl61_3 - terminalLocation not present in the MCDM message');
+    }
+    var f48sf61_1;
+    try {
+        f48sf61_1 = txn.getMcdmSchemaData().getMcdmposDataExtendedCondCodes().getPartialApprlTermSuppInd() === 1 ? true : false;
```
Comment (dianavicol) [inline ISS/develop/scripts/Build Request Message.js:248]: Same here, `isPartialApprlTermSuppInd()` could be used

File: ISS/develop/scripts/Build Request Message.js [hunk 3]
```diff
@@ -187,8 +214,43 @@ function populateTerminalDataObj(request) {
...
+        terminalDataObj.terminalSupportIndicator = f48sf61_2;
+    } catch (ex) {
+        log.debug('DE48sf61_2 -  terminalSupportIndicator not present in the MCDM message');
+    }
+
+    try {
+        var sf61_3 = txn.getMcdmData().getMcdmposdata().getPosterminallocation();
+        terminalDataObj.terminalLocation = sf61_3;
+    } catch (ex) {
+        log.debug('DEl61_3 - terminalLocation not present in the MCDM message');
```
Comment (dianavicol) [inline ISS/develop/scripts/Build Request Message.js:244]: Do we need the log message if the field is not mandatory?

File: ISS/develop/scripts/Build Request Message.js [hunk 4]
```diff
@@ -187,8 +214,43 @@ function populateTerminalDataObj(request) {
...
+        log.debug('DE61sf7 - terminalTrxStatus not present in the MCDM message');
+    }
+
+    try {
+        var sf61_10 = txn.getMcdmData().getMcdmposdata().getCardholderactivatedterminallevel();
+        terminalDataObj.terminalCATLevel = sf61_10;
+    } catch (ex) {
+        log.debug('DE61sf10 - terminalCATLevel not present in the MCDM message');
+    }
+
```
Comment (dianavicol) [inline ISS/develop/scripts/Build Request Message.js:232]: Could you extract `txn.getMcdmData().getMcdmposdata()` in a var and use it instead. I would be easier to read the code. Same for `getMcdmposDataExtendedCondCodes()` from below.

**Review by dianavicol — CHANGES_REQUESTED** [review]
Looks good, just a few points

File: ISS/develop/scripts/Process Response Message.js [hunk 5]
```diff
@@ -56,53 +56,67 @@ function start() {
...
+            error = 'Error occurred when populating billing amount partial: ' + ex.toString();
+            declineAndStoreReversal(ex);
+            return;
         }
 
-        if (!txn.isApproved()) {
-            txn.addFailReason('Host decline (' + response.responseCode + ')');
+    } else {
+        //check if it's partial approval
+        if (response.commonResponse.responseCode === 10 || response.commonResponse.responseCode === 87) {
```
Comment (dianavicol) [inline ISS/develop/scripts/Process Response Message.js:107]: As discussed, the response code from txn object should be checked.

File: ISS/develop/scripts/Build Request Message.js (hunk 4 above)
Comment (fsemerdzhieva) [reply ISS/develop/scripts/Build Request Message.js:232]: yeah but then I will have to take this outside my try and catch statements and mcdmData is not always initialised

Comment (dianavicol) [reply ISS/develop/scripts/Build Request Message.js]: Yeah, you could check it at the beginning, and if it is not initialised, you throw an exception, as it is needed for setting up the required request fields.

**Comment by dianavicol** [comment]
Thanks! 👍

**Pull request merged**